### PR TITLE
use block cache and reuse buffer to reduce memory allocations

### PR DESCRIPTION
**Pull request by hslam**
* Use block cache
* Reuse buffer

### CODE REVIEW

File: buffer/buffer.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	if pageSize < 1 {
+		b.pageSize = 1
+	} else {
+		b.pageSize = pageSize
+	}
+	return b
+}
+
+// AssignPool assigns a fixed size bytes pool with the given size.
+func (b *Buffers) AssignPool(size int) (p *Pool) {
```
Comment (coocood) [inline buffer/buffer.go:58]: Can we make it unexported?

File: table/sstable/table_file.go [hunk 2]
```diff
@@ -170,19 +175,31 @@ func NewLocalFile(filename string, blockCache, indexCache *cache.Cache) (TableFi
...
+		}
+		return blk, length, nil
 	})
 	if err != nil {
-		return nil, err
+		return &block{}, err
+	}
+	b := blk.(*block)
+	if ok := b.add(); !ok {
+		return &block{}, errors.Errorf("block is evicted")
```
Comment (coocood) [inline table/sstable/table_file.go:200]: we can retry for this error, so the outer user don't need to handle this error.

File: buffer/buffer.go (hunk 1 above)
Comment (hslam) [reply buffer/buffer.go:58]: There is a time cost to assign a bytes pool. 
When we need to put the buffer immediately after using it.
```
p := buffer.AssignPool(size)
buf = p.GetBuffer(size)
p.PutBuffer(buf)
```
faster than
```
buf = buffer.GetBuffer(size)
buffer.PutBuffer(buf)
```

File: table/sstable/table_file.go (hunk 2 above)
Comment (hslam) [reply table/sstable/table_file.go]: fixed

**Comment by coocood** [comment]
LGTM

**Pull request merged**